### ISSUE TITLE
Maintain optimiser object during training

### ISSUE THREAD

**Issue by tom-andersson** [tom-andersson/deepsensor#47]
Currently, `deepsensor.train.train_epoch` re-instantiates the Adam optimiser object at the start of each epoch. This is sub-optimal because it loses the momentum information gathered from previous training epochs. There should be a new training *class* which keeps track of the optimiser object (and thus momentum), which should speed up and stabilise training.

**Comment by tom-andersson** [comment]
Closed by https://github.com/tom-andersson/deepsensor/commit/ba04343accf2ca3bf5b641fa50915b83dbe72fe4 with a new `deepsensor.train.train.Trainer` class. I've kept `train_epoch` for backward compatibility -- users can still just user `train_epoch`, but are advised to instead use `Trainer` like below:
```
from deepsensor.train.train import Trainer
trainer = Trainer(model, lr=5e-5)
for epoch in tqdm(range(n_epochs)):
    batch_losses = trainer(train_tasks, batch_size=batch_size)
```

I've tested this by training a ConvGNP on the default xarray air temperature reanalysis dataset. Using `Trainer` improves the training loss and also decreases the spread of the loss after the initial epochs by ~20%. This is just based on one random seed, but at least the results are what we'd expect theoretically, so I'm happy with this.

![losses_comparison](https://github.com/tom-andersson/deepsensor/assets/26459412/42d9824b-32a5-4142-93c9-de13fedbd3d0)
